src/ranged.cpp

```diff
@@ -361,11 +361,19 @@ void player::fire_gun( const tripoint &targ_arg, bool burst )
         debugmsg("game::fire() - num_shots = 0!");
     }
 
+    bool mod_ups = false;
+    int mod_ups_drain = 0;
+    for( auto &mod : used_weapon->contents ){
+        if( mod.type->gunmod->ups_charges > 0 ){
+            mod_ups = true;
+            mod_ups_drain += mod.type->gunmod->ups_charges;
+        }
+    }
     int ups_drain = 0;
     int adv_ups_drain = 0;
     int bio_power_drain = 0;
-    if( used_weapon->type->gun->ups_charges > 0 ) {
-        ups_drain = used_weapon->type->gun->ups_charges;
+    if( used_weapon->type->gun->ups_charges > 0 || mod_ups ) {
+        ups_drain = used_weapon->type->gun->ups_charges + mod_ups_drain;
```

Comment:
The whole calculation of `ups_drain` is repeated in game..cpp, it should be in a separate function, maybe `item::get_gun_ups_drain` or similar.

Otherwise it looks fine.